cpp/src/decisiontree/batched-levelalgo/quantiles.cuh

```diff
@@ -17,74 +17,89 @@
 #pragma once
 
 #include <cub/cub.cuh>
+#include <fstream>
+#include <iostream>
 #include <memory>
+#include <omp.h>
 #include <raft/cuda_utils.cuh>
 #include <raft/handle.hpp>
 #include <rmm/device_uvector.hpp>
+#include <rmm/device_vector.hpp>
 #include <rmm/exec_policy.hpp>
 #include <thrust/fill.h>
 
 #include <raft/common/nvtx.hpp>
 
+#include "quantiles.h"
+
 namespace ML {
 namespace DT {
 
 template <typename T>
-__global__ void computeQuantilesSorted(T* quantiles,
-                                       const int n_bins,
-                                       const T* sorted_data,
-                                       const int length);
+__global__ void computeQuantilesBatchSorted(
+  T* quantiles, int* useful_nbins, const T* sorted_data, const int n_bins, const int length);
 
 template <typename T>
-std::shared_ptr<rmm::device_uvector<T>> computeQuantiles(
-  int n_bins, const T* data, int n_rows, int n_cols, const raft::handle_t& handle)
+void computeQuantiles(Quantiles<T, int>& quantiles,
+                      int n_bins,
+                      const T* data,
+                      int n_rows,
+                      int n_cols,
+                      int n_streams,
+                      const raft::handle_t& handle)
 {
-  auto quantiles = std::make_shared<rmm::device_uvector<T>>(n_bins * n_cols, handle.get_stream());
-  thrust::fill(rmm::exec_policy(handle.get_stream()),
-               quantiles->begin(),
-               quantiles->begin() + n_bins * n_cols,
-               0.0);
-  // Determine temporary device storage requirements
-  size_t temp_storage_bytes = 0;
+  raft::common::nvtx::push_range("computeQuantiles");
 
-  rmm::device_uvector<T> single_column_sorted(n_rows, handle.get_stream());
+  int prllsm = n_streams;  // the parallism to be used stream-wise and omp-thread-wise
+  size_t temp_storage_bytes = 0;
+  rmm::device_uvector<T> all_column_sorted(n_cols * n_rows, handle.get_stream());
```

Comment:
This is creating a full copy of the input data right? So the peak memory usage is potentially much greater. It's probably not worth doubling peak memory usage to get a small performance improvement sometimes.

Reply:
yep :( I think I'll revert the stream-parallelism as the gains probably don't justify the peak memory usage